README.md
# Recurly-nodejs

This is a fork of original `node-recurly` library by [Rob Righter](https://github.com/robrighter) for the recurly recurring billing service.

This library is intended to follow very closely the recurly documentation found at: [Recurly Docs](http://docs.recurly.com/)


## Installation
```
npm install recurly-nodejs --save
```

Add a config file to your project that has contents similar to:
```javascript
module.exports = {
  API_KEY: 'secret',
  HOST: 'recurly.com',
  PORT: 443,
  SUBDOMAIN:    '[your_account]',
  ENVIRONMENT:  'sandbox',
  DEBUG: false,
  API_VERSION: 2.7 // optional
};
```

## Usage


#### Using callbacks

```javascript
var Recurly = require('recurly-nodejs');
var recurly = new Recurly(require('./config'));
recurly.accounts.get('account_code_123', function (errResponse, response) {
})
```

#### Or a promises based version

```javascript
var Recurly = require('recurly-nodejs/promise');
var recurly = new Recurly(require('./config'));
recurly.accounts.get('account_code_123')
  .then(function (response) {})
  .catch(function (errorResponse) {})
```

For convenience the original callback version of every method is available with Callback suffix
```javascript
var callback = function () {};

var filter = {
  state: 'active'
};

recurly.subscriptions.listByAccountCallback('account_code_123', callback, filter)
```

### Accounts
> http://docs.recurly.com/api/accounts

```javascript
recurly.accounts.list(callback, filter)
recurly.accounts.create(details, callback)
recurly.accounts.update(accountcode, details, callback)
recurly.accounts.get(accountcode, callback)
recurly.accounts.close(accountcode, callback)
recurly.accounts.reopen(accountcode, callback)
recurly.accounts.notes(accountcode, callback)
```

### Billing Information
> http://docs.recurly.com/api/billing-info

```javascript
recurly.billingInfo.update(accountcode, details, callback)
recurly.billingInfo.get(accountcode, callback)
recurly.billingInfo.remove(accountcode, callback)
```

### Adjustments
> http://docs.recurly.com/api/adjustments
```javascript
recurly.adjustments.list(accountcode, callback)
recurly.adjustments.get(uuid, callback)
recurly.adjustments.create(accountcode, details, callback)
recurly.adjustments.remove(uuid, callback)
```

### Coupons
> http://docs.recurly.com/api/coupons

```javascript
recurly.coupons.list(callback, filter)
recurly.coupons.get(couponcode, callback)
recurly.coupons.create(details, callback)
recurly.coupons.deactivate(couponcode, callback)
```

### Coupon Redemption
> http://docs.recurly.com/api/coupons/coupon-redemption

```javascript
recurly.couponRedemption.redeem(couponcode, details, callback)
recurly.couponRedemption.get(accountcode, callback)
recurly.couponRedemption.remove(accountcode, callback)
recurly.couponRedemption.getByInvoice(invoicenumber, callback) // Deprecated
recurly.couponRedemption.getAllByInvoice(invoicenumber, callback)
```

### Invoices
> http://docs.recurly.com/api/invoices

```javascript
recurly.invoices.list(callback, filter)
recurly.invoices.listByAccount(accountcode, callback, filter)
recurly.invoices.get(invoicenumber, callback)
recurly.invoices.create(accountcode, details, callback)
recurly.invoices.preview(accountcode, callback)
recurly.invoices.refundLineItems(invoicenumber, details, callback)
recurly.invoices.refundOpenAmount(invoicenumber, details, callback)
recurly.invoices.markSuccessful(invoicenumber, callback)
recurly.invoices.markFailed(invoicenumber, callback)
recurly.invoices.enterOfflinePayment(invoicenumber, details, callback)
```

Special pdf method - callback will contain a pdf document as `Buffer`
You should send the buffer to the client with content type of `application/pdf`

```javascript
recurly.invoices.retrievePdf(invoicenumber, details, callback)
```

### Subscriptions
> http://docs.recurly.com/api/subscriptions

```javascript
recurly.subscriptions.list(callback, filter)
recurly.subscriptions.listByAccount(accountcode, callback, filter)
recurly.subscriptions.get(uuid, callback)
recurly.subscriptions.create(details, callback)
recurly.subscriptions.preview(details, callback)
recurly.subscriptions.update(uuid, details, callback)
recurly.subscriptions.updateNotes(uuid, details, callback)
recurly.subscriptions.updatePreview(uuid, details, callback)
recurly.subscriptions.cancel(uuid, callback)
recurly.subscriptions.reactivate(uuid, callback)
recurly.subscriptions.terminate(uuid, refundType, callback)
recurly.subscriptions.postpone(uuid, nextRenewalDate, callback)
```

### Subscription Plans
> http://docs.recurly.com/api/plans

```javascript
recurly.plans.list(callback, filter)
recurly.plans.get(plancode, callback)
recurly.plans.create(details, callback)
recurly.plans.update(plancode, details, callback)
recurly.plans.remove(plancode, callback)
```

### Plan Add-ons
> http://docs.recurly.com/api/plans/add-ons

```javascript
recurly.planAddons.list(plancode, callback, filter)
recurly.planAddons.get(plancode, addoncode, callback)
recurly.planAddons.create(plancode, details, callback)
recurly.planAddons.update(plancode, addoncode, details, callback)
recurly.planAddons.remove(plancode, addoncode, callback)
```

### Purchases
> https://dev.recurly.com/docs/create-purchase

```javascript
recurly.purchases.create(details, callback)
```

The purchase endpoint requires API version v2.6. Creating multiple subscriptions requires
API v2.8, and some extra feature flags enabled. Contact Recurly support for more details.

### Transactions
> http://docs.recurly.com/api/transactions

```javascript
recurly.transactions.list(callback, filter)
recurly.transactions.listByAccount(accountcode, callback, filter)
recurly.transactions.get(id, callback)
recurly.transactions.create(details, callback)
recurly.transactions.refund(id, callback, amount)
```

### Usage Records
> https://dev.recurly.com/docs/usage-record-object

```javascript
recurly.usageRecords.list(subscription_uuid, add_on_code, callback, filter)
recurly.usageRecords.lookup(subscription_uuid, add_on_code, usage_id, callback)
recurly.usageRecords.log(subscription_uuid, add_on_code, details, callback)
recurly.usageRecords.update(subscription_uuid, add_on_code, usage_id, details, callback)
recurly.usageRecords.delete(subscription_uuid, add_on_code, usage_id, callback)
```

### Custom API calls

```javascript
var options = {
  url: '/v2/accounts/' + account_code + '/invoices',
  method: 'POST',
  headers: {
    "My-Custom_header": "Value"
  },
  bodyRoot: 'invoice', // xml root element
  body: {} // content to convert to xml using js2xmlparser
};

recurly.api(options, callback)
```


CHANGELOG.md
# Change Log

## v3.3.0 - 2019-10-09

* added additional config options

## v3.2.0 - 2019-04-12

* add support for filters to subscriptions.listByAccount #45 [thomas-hilaire]

## v3.1.1 - 2019-02-10

* add missed getAllByInvoice method #44 [everdance]

## v3.1.0 - 2019-01-12

* get all invoice redemptions #43 [everdance]

## v3.0.0 - 2018-07-14

* add content-type header #33 [MrBar42]
* For all response status over 400, parse the response body as XML. #21 [bluecollarcoder]
* chore: adds Vulnerabilities badge #40 [Luke Watts]

## 2.7.0 - 2018-06-12

* Remove trailing comma from the invoices object [Thodoris Greasidis]
* fix: adds alias to match recurlyApi docs #37 [Luke Watts]

## 2.6.0 - 2018-04-09

* Add support for creating purchases #36 [Tim Perry]



lib/client.js
'use strict';

var https = require('https');
var pjson = require('../package.json');
var Xml2js = require('xml2js');

exports.create = function(config) {
	var host = config.SUBDOMAIN ? `${config.SUBDOMAIN}.${config.HOST}` : config.HOST;
	var port = config.PORT || 443;
	var rejectUnauthorized = typeof config.REJECT_UNAUTHORIZED === 'boolean' ? config.REJECT_UNAUTHORIZED : true;
	var parser = new Xml2js.Parser(config.XML2JS_OPTIONS || { explicitArray: false });

	return {
		request: function(route, callback, data) {
			var path = route[0];
			var method = route[1];
			var headers = route[2];
			var that = this;
			var options = {
				host,
				port,
				method,
				path,
				rejectUnauthorized,
				headers: {
					Authorization: 'Basic ' + new Buffer(config.API_KEY).toString('base64'),
					Accept: 'application/xml',
					'Content-Length': data ? Buffer.byteLength(data, 'utf8') : 0,
					'Content-Type': 'application/xml; charset=utf-8',
					'User-Agent': 'recurly-nodejs/' + pjson.version,
				},
			};

			if (config.API_VERSION) {
				options.headers['X-Api-Version'] = config.API_VERSION;
			}

			if (headers) {
				for (var p in headers) {
					if (!headers.hasOwnProperty(p)) continue;

					options.headers[p] = headers[p];
				}
			}

			if (method.toLowerCase() === 'post' || method.toLowerCase() === 'put') {
				options.headers['Content-Type'] = 'application/xml';
				that.debug(data);
			}
			that.debug(options);
			var req = https.request(options, function(res) {
				var responsedata = [];
				res.on('data', function(d) {
					responsedata.push(d);
				});
				res.on('end', function() {
					responsedata = Buffer.concat(responsedata);
					that.debug('Response is: ' + res.statusCode);
					that.debug(responsedata);
					try {
						// 200–299 success
						if (res.statusCode >= 200 && res.statusCode <= 299) {
							if (responsedata === '') {
								return _cb(res);
							} else if (options.headers.Accept !== 'application/xml') {
								return _cb(res, null, responsedata);
							}
							return parser.parseString(responsedata, function(err, result) {
								return _cb(res, null, result);
							});
						}
						// 400–499 client request errors
						// 500–599 server errors
						if ([404, 412, 422, 500].indexOf(res.statusCode) !== -1) {
							return parser.parseString(responsedata, function(err, result) {
								return _cb(res, result);
							});
						}
						if (res.statusCode >= 400) {
							return parser.parseString(responsedata, function(err, result) {
								return _cb(res, result);
							});
						}
					} catch (e) {
						return _cb(null, e);
					}
				});
			});
			if (data) {
				req.write(data);
			}
			req.end();
			req.on('error', function(e) {
				return _cb(null, e);
			});
			// fallback for backward compatibility
			function _cb(res, err, data) {
				// callback objects acquired from parent scope
				if (typeof callback === 'undefined') {
					throw new Error('Missing argument: callback function');
				}
				if (typeof callback !== 'function') {
					throw new Error('Callback should be a function');
				}
				if (callback.length === 2) {
					if (err) {
						return callback(_wrapResponse(res, err));
					}
					return callback(null, _wrapResponse(res, data));
				}
				// backward compatibility for not node.js style callbacks
				// TBD: skip in next version?
				else if (callback.length === 1) {
					var toreturn = { status: 'ok', data: '', headers: res ? res.headers : null };
					if (err) {
						toreturn.status = 'error';
						if (!res || err === Error || err instanceof Error) {
							toreturn.description = err;
						} else if (res.statusCode >= 400) {
							toreturn.data = res.statusCode;
							toreturn.additional = err;
						} else {
							toreturn.data = err;
						}
						return callback(toreturn);
					}
					toreturn.data = data;
					toreturn.description = res.statusCode;
					return callback(toreturn);
				}
			}

			function _wrapResponse(res, data) {
				if (!res) {
					return { data: data && data.stack ? data.stack : data };
				}
				return {
					statusCode: res.statusCode,
					headers: res.headers,
					data: data,
				};
			}
		},

		debug: function(s) {
			if (config.DEBUG) {
				console.log(s);
			}
		},
	};
};
